sdk/data_kusto/src/operations/query.rs

```diff
@@ -0,0 +1,174 @@
+use crate::client::KustoClient;
+use crate::request_options::*;
+use azure_core::prelude::*;
+use azure_core::setters;
+use azure_core::{
+    collect_pinned_stream,
+    headers::{add_mandatory_header2, add_optional_header2},
+    Response as HttpResponse,
+};
+use futures::future::BoxFuture;
+
+/// A future of a delete file response
+type ExecuteQuery = BoxFuture<'static, crate::error::Result<KustoResponseDataSetV2>>;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct QueryBody {
+    db: String,
+    csl: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct ExecuteQueryBuilder {
+    client: KustoClient,
+    database: String,
+    query: String,
+    client_request_id: Option<ClientRequestId>,
+    context: Context,
+}
+
+impl ExecuteQueryBuilder {
+    pub(crate) fn new(
+        client: KustoClient,
+        database: String,
+        query: String,
+        context: Context,
+    ) -> Self {
+        Self {
+            client,
+            database,
+            query,
+            client_request_id: None,
+            context,
+        }
+    }
+
+    setters! {
+        client_request_id: ClientRequestId => Some(client_request_id),
+        query: String => query,
+        database: String => database,
+        context: Context => context,
+    }
+
+    pub fn into_future(self) -> ExecuteQuery {
```

Comment:
I think we can completely avoid Box pin in this method and also simplify it:


```rust
pub async fn into_future(self) -> crate::error::Result<KustoResponseDataSetV2> {
        let url = self.client.query_url();
        let mut request = self.client.prepare_request(url, http::Method::POST);

        add_mandatory_header2(
            &ContentType::new("application/json; charset=utf-8"),
            &mut request,
        )
            .unwrap();
        add_mandatory_header2(&Accept::new("application/json"), &mut request).unwrap();
        //add_mandatory_header2(&AcceptEncoding::new("gzip,deflate"), &mut request).unwrap();
        add_optional_header2(&self.client_request_id, &mut request).unwrap();

        let body = QueryBody {
            db: self.database,
            csl: self.query,
        };
        request.set_body(bytes::Bytes::from(serde_json::to_string(&body)?).into());

        let mut context = self.context;

        let response = self
            .client
            .pipeline()
            .send(&mut context, &mut request)
            .await?;

        KustoResponseDataSetV2::try_from(response).await
    }
```

Reply:
Tried to adjust the `into_future` method to the latest approach from the cosmos crate and also added some more optional request options to illustrate the usage of the request builder. Also added the implementation for the so far only in nightly `IntoFuture` trait to show how the explicit `inio_future` call will dissapear once that feature lands in stable.

Reply:
You still left the box::pin and copying which again I don't think are needed here

Reply:
We had one issue in the past, that i think will re-appear if we drop that. Essentially, when used in API we may want to return the future rather then evaluating the future in the API itself, and then turning the response into a future again. Main reason being that if we don't copy the client and box/pin it, the lifetime of the future is directly tied to the lifetime of the client itself, which sometimes significantly limits usability in async contexts. However I am not entirely sure that I fully understand all implications - I think @rylev and @yoshuawuyts did the majority of the designs around this topic. 

a related issues (i think):
- #164

Reply:
Trying to understand this. Aside from the `Box::pin`, if we do not clone the client and do the `async move`, my understanding was we would again run into the situation, where the futures and clients lifetimes would be correlated, and could not have a function that return the future directly for use in downstream functions, but rather have to await the future and then wrap the returned value into a future again. We ran into a situation like this with the `list_blobs` function from storage in the delta-rs crate, and I wanted to understand the actual root cause, since we have yet to clean that um in delta.